src/plugins/intel_cpu/src/nodes/kernels/registers_pool.hpp

```diff
@@ -0,0 +1,260 @@
+            return idx;
+        } else if (std::is_same<TReg, Xbyak::Reg64>::value || std::is_same<TReg, Xbyak::Reg32>::value) {
+            auto idx = generalSet.getUnused();
+            generalSet.setAsUsed(idx);
+            return idx;
+        } else if (std::is_same<TReg, Xbyak::Opmask>::value) {
+            return getFreeOpmask();
+        }
+    }
+
+    template<typename TReg>
+    void returnToPool(TReg reg) {
+        if (std::is_same<TReg, Xbyak::Xmm>::value || std::is_same<TReg, Xbyak::Ymm>::value || std::is_same<TReg, Xbyak::Zmm>::value) {
+            simdSet.setAsUnused(reg.getIdx());
+        } else if (std::is_same<TReg, Xbyak::Reg64>::value || std::is_same<TReg, Xbyak::Reg32>::value) {
+            generalSet.setAsUnused(reg.getIdx());
+        } else if (std::is_same<TReg, Xbyak::Opmask>::value) {
+            returnOpmaskToPool(reg.getIdx());
+        }
+    }
+
+    void checkUniqueAndUpdate(bool isCtor = true) {
+        static thread_local bool isCreated = false;
+        if (isCtor) {
+            if (isCreated) {
+                IE_THROW() << "There should be only one instance of RegistersPool per thread";
+            }
+            isCreated = true;
+        } else {
+            isCreated = false;
+        }
+    }
+
+    PhysicalSet generalSet {16};
+    PhysicalSet simdSet;
+};
+
+template<typename TReg>
+RegistersPool::Reg<TReg>::Reg(const RegistersPool::Ptr& regPool) {
+    static_assert(std::is_same<TReg, Xbyak::Xmm>::value || std::is_same<TReg, Xbyak::Ymm>::value || std::is_same<TReg, Xbyak::Zmm>::value ||
+                  std::is_same<TReg, Xbyak::Reg64>::value || std::is_same<TReg, Xbyak::Reg32>::value || std::is_same<TReg, Xbyak::Opmask>::value,
+                  "The type is not supported for the RegistersPool::Reg template");
+    reg = TReg(regPool->template getFree<TReg>());
+    this->regPool = regPool;
+}
+
+template <x64::cpu_isa_t isa>
+class IsaRegistersPool : public RegistersPool {
+public:
+    IsaRegistersPool(std::initializer_list<Xbyak::Reg> regsToExclude) : RegistersPool(regsToExclude, x64::cpu_isa_traits<isa>::n_vregs) {}
+};
+
+template <>
+class IsaRegistersPool<x64::avx512_core> : public RegistersPool {
+public:
+    IsaRegistersPool(std::initializer_list<Xbyak::Reg> regsToExclude)
+            : RegistersPool(regsToExclude, x64::cpu_isa_traits<x64::avx512_core>::n_vregs) {}
+
+    int getFreeOpmask() override {
+        auto idx = opmaskSet.getUnused();
+        opmaskSet.setAsUsed(idx);
+        return idx;
+    }
+
+    void returnOpmaskToPool(int idx) override {
+        opmaskSet.setAsUnused(idx);
+    }
+
+    void excludeOpmask(const Xbyak::Opmask& reg) override {
+        opmaskSet.exclude(reg);
+    }
+
+    size_t countUnusedOpmask() const override {
+        return opmaskSet.countUnused();
+    }
+
+protected:
+    PhysicalSet opmaskSet {8};
+};
+
+template <x64::cpu_isa_t isa>
+RegistersPool::Ptr RegistersPool::create(std::initializer_list<Xbyak::Reg> regsToExclude) {
```

Comment:
As we discussed internally this method actually not needed, because `Reg` requires `std::shared_ptr<>`.
Also usually for factory method it is better to use `perfect forwarding`:
```cpp
template <x64::cpu_isa_t isa>
RegistersPool::Ptr RegistersPool::create(const TArgs&&... args) {
    return std::make_shared<IsaRegistersPool<isa>>(std::forward<TArgs>(args));
}
```
Because you will not need to rewrite signature of `RegistersPool::Ptr RegistersPool::create()` in future